trie/iter.py

```diff
@@ -0,0 +1,93 @@
+from trie.constants import (
+    NODE_TYPE_BLANK,
+    NODE_TYPE_LEAF,
+    NODE_TYPE_EXTENSION,
+    NODE_TYPE_BRANCH,
+)
+from trie.exceptions import (
+    InvalidNode,
+)
+from trie.utils.nibbles import (
+    bytes_to_nibbles,
+    nibbles_to_bytes,
+    remove_nibbles_terminator,
+)
+from trie.utils.nodes import (
+    get_node_type,
+    extract_key,
+    key_starts_with,
+)
+
+
+# FIXME: This is probably very inneficient and the code itself is quite convoluted. We should
+# either look into refactoring or rewriting it from scratch eventually.
+class NodeIterator:
+    """Iterate over all nodes of a trie, ensuring its consistency."""
+
+    def __init__(self, trie):
+        self.trie = trie
+
+    def next(self, key):
+        key = bytes_to_nibbles(key)
+        nibbles = self._iter(self.trie.root_node, key)
+        if nibbles is None:
+            return None
+        return nibbles_to_bytes(remove_nibbles_terminator(nibbles))
+
+    def _get_next(self, node):
+        node_type = get_node_type(node)
+        if node_type == NODE_TYPE_BLANK:
+            return None
+        elif node_type == NODE_TYPE_LEAF:
+            curr_key = extract_key(node)
+            return curr_key
+        elif node_type == NODE_TYPE_EXTENSION:
+            curr_key = extract_key(node)
+            sub_node = self.trie._get_node(node[1])
```

Comment:
It'd be nice if we could avoid using the internal `trie._get_node`.  Maybe instead we could add a new method on the trie class for retrieving *raw* nodes or something?  `Trie.get_raw_node`?

Reply:
Well, isn't that just a matter of renaming it to `trie.get_node()` (I'd prefer that over `get_raw_node()` as the latter kind of suggests it'd return the RLP-encoded node)?

Reply:
My mistake.  `trie.get_node()` is fine.  For some reason I thought there was already a `get_node`, hence the new name suggestion.